### DISCUSSION TITLE
Show Originals: UI Ideas

### DISCUSSION THREAD

**Discussion by marcustyphoon** [AprilSylph/XKit-Rewritten#461] [Suggestion Box]
Can you self-assign a suggestion discussion? Whatever.

Anyways, I had the idea to implement a simple visual indicator that Show Originals is active on a timeline (using a `::before` css element):

<img width="700" src="https://user-images.githubusercontent.com/8336245/153102730-4303f933-751d-42ae-905d-f49ab5653292.png">

(Not in love with the CSS there, but I haven't come up with anything better.)

But in looking around the site for UI elements to take inspiration from, I stumbled upon the timeline options element in the explore tab, and realized that this could be quite slick (and work just fine in the mobile view):

<img width="700" src="https://user-images.githubusercontent.com/8336245/153103012-6394a2f4-85d3-41ad-8cbe-2452bf7a0d52.png">
 
Would you be interested in either of these, April? Naturally, the second one would be more code, though it would be pretty much entirely implementing the options element itself (I would simply remove the timeline filter from processPosts and change the css to `.xkit-show-originals-timeline .xkit-show-originals-hidden article { display: none; }`.

Yes, if I implement the second one I will wind up undoing a ~~not-significant~~ *not-insignificant* amount of my own PR. Ssh.

**Comment by AprilSylph** [comment]
The second screenshot looks amazing. I say go for it! UX improvement justifies the increase in code complexity.

**Comment by marcustyphoon** [comment]
Hm, the big question I have, then, is: how should your selection of originals-or-all-posts be handled? I can think of...

- There are still checkboxes in the settings panel; the page element toggles them. I think this would be the first time we do that specific effect, not that I can think of anything fundamentally difficult/problematic about it. Would implement an onStorageChanged function to avoid reloading the script on toggle.
- There are no checkboxes in the settings panel; your most recent selection is quietly remembered just like the blog selector functionality we just implemented in quick reblog.
- Just default to the originals tab. (Or even the all posts tab, though if you bothered to enable the script, it seems fairly unlikely you want that, I would guess).

I can think of lots of other implementation specific interesting things, but I'll get to those when I try it.

**Reply by AprilSylph** [reply]
Quietly remembering seems like the best idea to me.

**Comment by marcustyphoon** [comment]
Just a random observation: I find it really interesting implementing a oneoff stateful component in this codebase, since what XKit does is so skewed towards modifications rather than additions, so there aren't many similar pieces of code to aim for a consistent style with. It's like the wild west!

Fun to watch my brain repeatedly reinvent vDOM: "if you select one of the buttons, then the other button should become unselected... ah, so the rendered state would just be a strict equality check with the current selection. but then you need to update both elements on any change, so they both would have to subscribe... oh, look, it's Svelte. or, you could just go declarative and render them both... oh, look, it's React."

**Reply by AprilSylph** [reply]
Or you could use a single attribute on the parent element and style the buttons using that 🤔

**Reply by marcustyphoon** [reply]
Essentially like this? Indeed, should work great in a case like this!
```css
.stateOne .buttonOne { /* highlighted */ }
.stateTwo .buttonTwo { /* highlighted */ }
```
I found it really interesting how you implemented `reblogPost` in quick reblog, querying the data attributes on the buttons themselves; don't think I would have thought of that!

**Reply by AprilSylph** [reply]
Essentially! I love overusing `data-*` attributes.

```css
[data-active="true"] > .showOriginalsButton { /* highlighted */ }
[data-active="false"] > .showAllPostsButton { /* highlighted */ }
```

This way, you can easily manipulate the style of both buttons from each button's `parentNode.dataset`. Just try not to rely on reading it in JS if it's simple to avoid, since any value stored on it will get converted to a string: https://github.com/AprilSylph/XKit-Rewritten/blob/25319e18886d5baaada9311f54c3f810874f1c4f/src/scripts/quick_reblog.js#L216-L230

Also, fun fact I realised fully recently: since `dataset` and `classList` are objects, destructuring them from their element doesn't clone them or disassociate them from their elements, instead creating a new reference to them. This sounds obvious, but it means you can manipulate them post-destructuring and still affect the element, as I do in Quick Reblog here.
